app/core/models.py

```diff
@@ -366,17 +373,26 @@ def check_cancel_requested(self):
         status = SubmissionRun.objects.values_list("status", flat=True).get(pk=self.id)
         return status == Status.CANCEL_PENDING
 
-    def mark_for_cancel(self):
+    def mark_for_cancel(self, remote=False):
         with transaction.atomic():
-            SubmissionRun.objects.filter(pk=self.id).update(
-                status=Status.CANCEL_PENDING
-            )
-            Evaluation.objects.filter(
-                submission_run_id=self.id, status=Status.PENDING
-            ).update(status=Status.CANCELLED)
-            Evaluation.objects.filter(
-                submission_run_id=self.id, status=Status.RUNNING
-            ).update(status=Status.CANCEL_PENDING)
+            SubmissionRun.objects.select_for_update().filter(pk=self.id)
+            submission_run = SubmissionRun.objects.get(pk=self.id)
```

Comment:
```suggestion
            submission_run = SubmissionRun.objects.select_for_update().get(pk=self.id)
```
but do the evaluation select_related first to lock all the rows